### ISSUE TITLE
Too Many Threads

### ISSUE THREAD

**Issue by Maithem** [CorfuDB/CorfuDB#989]
The server is creating too many handler threads to service requests. ForkJoinPool uses the number of cpus to set the parallelisim, but that number doesn't seem to be strictly enforced (because of the fork join pool implementation) . This can be illustrated by setting the parallelisim to 4 and creating 20 writer threads. The server will create a new thread per write request and ForkJoin pool will spawn more than 4 threads at a time. 

@no2chem since all write operations are blocking, do we really want this behavior: "unbounded" growth of threads?

**Comment by no2chem** [comment]
1) When is this a problem
2) Is the # of threads really unbounded? A FJ pool will only create a new thread if there aren't any threads available (probably the available threads are waiting for I/O).

**Comment by Maithem** [comment]
1) Possible starvation issues, fundamentally incoming requests can affect prior requests. Shouldn't the server reject requests that it cannot service?

2) From the implementation, I can see that there is an upper bound, but that bound is not the parallelism number that is passed to the constructor. A thread per request seems to be expensive. Also, in a FJ pool a new thread is created for every request (it seems), which seems expensive. In the profiler, I can see that the number of threads is almost equal to the number of concurrent client threads.

**Comment by no2chem** [comment]
1. Starvation issues? You mean running out of threads? Can you actually show some evidence for your speculation?  You would need 32k threads total to hit the limit, but perhaps you are referring to a memory issue? Again, you need some evidence for this claim.

Also, *fundamentally* previous requests will affect future requests. This is what happens when we have limited resources like I/O, no?

2. You seem to think threads are expensive. Can you ground your intuition in some evidence?

**Comment by Maithem** [comment]
tl;dr based on the latency/throughput stats that I have collected it seems like the forkjoin pool is slower than fixed sized thread pools, and my claim is the FJ pool is slower because it creates more threads than the parallelism parameter (which is set to the # of CPUs) and it because it keeps recreating threads.

Its more of an observation based on my perf analysis than pure speculation. I ran the experiment 3 days ago, but I just re-ran it to gather some numbers for you. In this test I run both the client and the server (in-memory) on the same machine. I ran two tests. One on a server that has a forkjoin pool and the other on a fixed sized pool. In each test I start with 1 writer and double the number of threads until it reaches 64 writer threads.

The data below shows data concerning latency for a write request.

Forkjoin, 1 writer thread

#[Mean    =         47.8, StdDeviation   =        107.6]
#[Max     =       6081.7, Total count    =         5000]
#[Buckets =           38, SubBuckets     =           32]

Forkjoin, 4 writer threads
#[Mean    =         50.0, StdDeviation   =         65.3]
#[Max     =       5872.0, Total count    =        19888]
#[Buckets =           38, SubBuckets     =           32]


Forkjoin, 8 writer threads
#[Mean    =         76.2, StdDeviation   =        132.1]
#[Max     =      10905.2, Total count    =        39830]
#[Buckets =           38, SubBuckets     =           32]


Forkjoin, 16 writer threads
#[Mean    =         95.3, StdDeviation   =        167.4]
#[Max     =      20132.7, Total count    =        79612]
#[Buckets =           38, SubBuckets     =           32]


Forkjoin, 32 writer threads
#[Mean    =        157.0, StdDeviation   =        227.4]
#[Max     =      25165.8, Total count    =       159274]
#[Buckets =           38, SubBuckets     =           32]


Forkjoin, 64 writer threads
#[Mean    =        336.9, StdDeviation   =        462.5]
#[Max     =      43620.8, Total count    =       318596]
#[Buckets =           38, SubBuckets     =           32]

fixedThreadPool, 1 writer thread
#[Mean    =         29.2, StdDeviation   =         79.9]
#[Max     =       5452.6, Total count    =         5000]
#[Buckets =           38, SubBuckets     =           32]

fixedThreadPool, 4 writer threads
#[Mean    =         43.6, StdDeviation   =         41.4]
#[Max     =       5242.9, Total count    =        19930]
#[Buckets =           38, SubBuckets     =           32]

fixedThreadPool, 8 writer threads
#[Mean    =         57.3, StdDeviation   =         54.7]
#[Max     =      33554.4, Total count    =        39770]
#[Buckets =           38, SubBuckets     =           32]

fixedThreadPool, 16 writer threads
#[Mean    =         65.1, StdDeviation   =         53.7]
#[Max     =      35232.2, Total count    =        79508]
#[Buckets =           38, SubBuckets     =           32]

fixedThreadPool, 32 writer threads
#[Mean    =        102.1, StdDeviation   =        105.8]
#[Max     =      43620.8, Total count    =       159182]
#[Buckets =           38, SubBuckets     =           32]

fixedThreadPool, 64 writer threads
#[Mean    =        192.6, StdDeviation   =        310.7]
#[Max     =      40265.3, Total count    =       318702]
#[Buckets =           38, SubBuckets     =           32]


The throughput stats also exhibit the same general trend. 

Anyways, the performance consequences (i.e. context switching time, memory pressure, thread creation cost) of too many threads is well documented online, so I'm not sure why you seem skeptical that this is a well founded problem problem.

**Comment by no2chem** [comment]
Well, in your post you claimed:

>Possible starvation issues, fundamentally incoming requests can affect prior requests, Shouldn't the server reject requests that it cannot service
You still haven't shown any *starvation* issues, nor does switching to a FixedThreadPool make the server reject requests it can't service.

> the performance consequences (i.e. context switching time, memory pressure, thread creation cost) of too many threads is well documented online
Sure. But what is too many here, and is having too many threads really the problem? Or does it just have to do with how ForkJoinPools manage threads? It sounds like you are trying to say that ForkJoinPools are bad. But clearly they are not universally bad, otherwise what would be the point in having them? Does it have to do with some way we are using them?

As for your benchmark, it does suggest a problem with the FJ pool with that workload. However, if the problem is too many threads, perhaps it would make sense to vary the number of client threads (how many were there?), as well as clients, and document client liveliness, which are all important factors to consider.

**Comment by Maithem** [comment]
The goal is to restrict the number of threads on the server in order to prevent thrashing. Whether this is done by rejecting requests, or using a fixed sized thread pool, that's besides the point.  

The reported numbers are latencies experienced by the client threads that are writing continuously. I didn't do any changes on the server except for changing the FJPool to a newFixedThreadPool, and maintained the the same "parallelism" number. Note that the "writer threads" are actually client threads.

I ran the server on an 8 core machine, which means that the server handler thread pool was initialized with parallelism set to 8. Then I ran the client with varying number of threads as shown in my last post.
For example, when I was running 64 writer threads, the server actually spawn ~50 threads in the FJPool,
and it seemed to recreate the threads continuously. 

The second set of numbers show that when you replace the FJPool with a fixed size thread pool (while maintaining the same parallelism number), you get better performance results. I also ran a third experiment where I restrict the number of write requests before they "hit" the FJPool, and the performance was almost commensurate with the fixed size thread pool.

Furthermore, the clients workload was simply continuously appending to the same stream. It is not a pathological case, the system is expected to handle write only loads. All writes are blocking, so the FJPool keeps creating more threads to maintain "parallelism" as the client writer threads increase. 

The aforementioned observations lead me to believe that the discrepancy between the two tests is caused by thread thrashing (context switching/Thread management costs), otherwise the FJPool should perform faster because it would decrease the queuing time that is experienced for tasks that run on a fixed thread pool.

**Comment by no2chem** [comment]
>The goal is to restrict the number of threads on the server in order to prevent thrashing.

We need to take this offline, but that's not the only way to prevent thrashing, if that's really what is happening.

> For example, when I was running 64 writer threads, the server actually spawn ~50 threads in the FJPool,

64 writer threads from the *client*? You weren't clear about this in your post about your results. With a single client?

> The aforementioned observations lead me to believe that the discrepancy between the two tests is caused by thread thrashing 

It's not clear yet, and thrashing is not the same as "/Thread management costs". 

You're talking about the ForkJoinPool in serverRouter right? Wouldn't replacing that pool with a FixedThreadPool cause the server to be unresponsive if we had more clients than threads in that pool? Keep in mind there are multiple services (layout, sequencer) on the same server....

**Comment by Maithem** [comment]
This issue potentially has two action items, leave it open for now.

**Comment by no2chem** [comment]
@Maithem - Looking further into this, I have several comments. 

First, you are right that the ForkJoinPool generates extra threads when blocked; up to 32k. You are also right that switching to a FixedThreadPool appears to help performance for a pure write workload. This is because for writes, spawning an additional thread is pointless: the write is going to be batched anyway, so the system is just increasing the work BatchWriter has to do, and BatchWriter won't complete until the current batch is done anyway.

However, if you perform a mixed workload, having more threads helps here, since if your FixedThreadPool unluckily schedules multiple writes to saturate the pool, calls to the sequencer will be blocked behind it. For this reason, a cachedThreadPool (which spawns unbounded threads) will do decently well here, especially on a workload that heavily queries the sequencer, spawning new threads if there are threads blocked on I/O. 

This circles back to the BatchWriter. It seems that how writes are handled now is that they block. The largest batch size is then equal to the size of the number threads, which isn't ideal. We should change the write to be non-blocking and schedule into the BatchWriter, this would eliminate the threading problem. I'd also note that if we did that, handlerWorkers executor becomes pretty pointless.

**Comment by no2chem** [comment]
Did some more experiments. Actually writing to disk matters a lot, because it increases the I/O blocking time incurred by each spawned thread, and this is when you can see the sequencer getting blocked by writer threads becoming a performance issue:

For a pure write workload (```insert``` on a CorfuTable, 16 byte keys/values, 16 client threads):
FixedThreadPool (2 threads): 3.66 ms/op avg
FixedThreadPool (4 threads): 3.24 ms/op avg
FixedThreadPool (8 threads): 3.16 ms/op avg
CachedThreadPool (unbounded threads):  2.74 ms/op avg
ForkJoinPool (32k threads):  3.11 ms/op avg

So a FixedThreadPool actually significantly degrades performance here.
